neqo-client/src/main.rs

```diff
@@ -391,63 +393,78 @@ fn get_output_file(
 fn process_loop(
     local_addr: &SocketAddr,
     socket: &UdpSocket,
+    poll: &Poll,
     client: &mut Http3Client,
     handler: &mut Handler,
 ) -> Res<neqo_http3::Http3State> {
     let buf = &mut [0u8; 2048];
+    let mut events = Events::with_capacity(1024);
+    let mut timeout: Option<Duration> = None;
     loop {
+        poll.poll(
+            &mut events,
+            if timeout.is_some() {
+                timeout
+            } else {
+                Some(Duration::from_millis(0))
+            },
+        )?;
+
+        'read: loop {
+            match socket.recv_from(&mut buf[..]) {
+                Err(ref err) => {
+                    if err.kind() == ErrorKind::WouldBlock || err.kind() == ErrorKind::Interrupted {
+                        break 'read;
+                    }
+
+                    eprintln!("UDP error: {}", err);
+                    exit(1)
+                }
+                Ok((sz, remote)) => {
+                    if sz == buf.len() {
+                        eprintln!("Received more than {} bytes", buf.len());
+                        break 'read;
+                    }
+                    if sz > 0 {
+                        let d = Datagram::new(remote, *local_addr, &buf[..sz]);
+                        client.process_input(d, Instant::now());
+                        handler.maybe_key_update(client)?;
```

Comment:
This still passes the packets one-by-one to neqo, so effectively it's doing exactly what we did before. I think what we want (also for GRO/GSO) it to be able to pass a `Vec<Datagram>` into neqo.

Reply:
I've tried to expose a new interface: `pub fn process_multiple_input(&mut self, dgrams: Vec<Datagram>, now: Instant)` to allow pass `Vec<Datagram>` to neqo, but I am not sure if this is enough.

Reply:
Actually, I think we want an array here and not a `Vec`, because IIRC there is some way to run the AEAD over multiple consecutive packets in a memory buffer rather than one-by-one. We'd need to pick a sane maximum array length, probably based on whatever the GSO/GRO limit is at the moment.